### ISSUE TITLE
Error message with `BUNDLE_GLOBAL_GEM_CACHE` enabled refers to incorrect file

### ISSUE THREAD

**Issue by captn3m0** [rubygems/rubygems#3863]
### Describe the problem as clearly as you can

If you have `BUNDLE_GLOBAL_GEM_CACHE` enabled in your bundler config, any corrupt gem messages will refer to the incorrect file path.

### Post system independent steps to reproduce the problem

1. Ensure that bundler global cache configuration is enabled
2. Have a corrupted download once, on any project on any gem
3. Try to install the gem using bundler anywhere else

The error message will look like the following:

`Bundler::GemspecError: Could not read gem at /local/pathvendor/bundle/ruby/2.7.0/cache/rake-13.0.1.gem.
It may be corrupted.`

### Which command did you run?

First we fake a corrupted gem download (which could happen for many reasons, but it is fairly regular for me) by creating an empty file:

`touch ~/.bundle/cache/gems/rubygems.org.443.29b0360b937aa4d161703e6160654e47/rake-13.0.1.gem`

I usually have 5-10 gem files in there that are empty.

Then I ran `bundle install`

### What were you expecting to happen?

Either of these 2:

1. `bundler` should notice that the `.gem` file is empty, and just ignore the cache
2. Barring that (if the downloaded gem is actually corrupt), the error message should point to `cache/gems/rubygems.org.443.29b0360b937aa4d161703e6160654e47/rake-13.0.1.gem`, instead of the non-existent `/home/nemo/projects/personal/bundler-gem-bug/vendor/bundle/ruby/2.7.0/cache/rake-13.0.1.gem`

### What actually happened?

```
Fetching gem metadata from https://rubygems.org/.
Bundler::GemspecError: Could not read gem at /home/nemo/projects/personal/bundler-gem-bug/vendor/bundle/ruby/2.7.0/cache/rake-13.0.1.gem.
It may be corrupted.
An error occurred while installing rake (13.0.1), and Bundler cannot continue.
Make sure that `gem install rake -v '13.0.1' --source 'https://rubygems.org/'` succeeds before bundling.

In Gemfile:
  rake
```

Notice that the error message points to a file that _does not exist_. I debugged this ages ago, and it was very quite hard to trace down (`strace` was what pointed me to the global gem cache).

## Environment

```
Bundler       2.1.4
  Platforms   ruby, x86_64-linux
Ruby          2.7.1p83 (2020-03-31 revision a0c7c23c9cec0d0ffcba012279cd652d28ad5bf3) [x86_64-linux]
  Full Path   /usr/bin/ruby
  Config Dir  /etc
RubyGems      3.1.3
  Gem Home    /usr/lib/ruby/gems/2.7.0
  Gem Path    /home/nemo/.gem/ruby/2.7.0:/usr/lib/ruby/gems/2.7.0
  User Home   /home/nemo
  User Path   /home/nemo/.gem/ruby/2.7.0
  Bin Dir     /usr/bin
Tools         
  Git         2.27.0
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2020-01-05
Git SHA           32a4159325
Released Version  true
```

## Bundler settings

```
path
  Set for the current user (/home/nemo/.bundle/config): "vendor/bundle"
#bundle_default_install_uses_path
  You have not configured a value for `#bundle_default_install_uses_path`
#bundle_allow_offline_install
  You have not configured a value for `#bundle_allow_offline_install`
# bundle_cache_all
  You have not configured a value for `# bundle_cache_all`
global_gem_cache
  Set for the current user (/home/nemo/.bundle/config): true
#bundle_gem.test
  You have not configured a value for `#bundle_gem.test`
#bundle_gem.mit
  You have not configured a value for `#bundle_gem.mit`
#bundle_build.nokogiri
  You have not configured a value for `#bundle_build.nokogiri`
#bundle_disable_shared_gems
  You have not configured a value for `#bundle_disable_shared_gems`
#bundle_gem.coc
  You have not configured a value for `#bundle_gem.coc`
```

### Gemfile

```ruby
source 'https://rubygems.org'

gem 'rake', '13.0.1'
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    rake (13.0.1)

PLATFORMS
  ruby

DEPENDENCIES
  rake (= 13.0.1)

BUNDLED WITH
   2.1.4
```

**Comment by deivid-rodriguez** [comment]
I completely removed this error message in https://github.com/rubygems/rubygems/pull/3774. I think that might've fixed this issue as a side effect and the underlying error most likely includes a correct path?

**Comment by captn3m0** [comment]
I applied the patch, and while the error message changes, the result is the same (FormatError pointing to the incorrect file path):

```
bundle install
Fetching gem metadata from https://rubygems.org/.
Gem::Package::FormatError: package metadata is missing in
/home/nemo/projects/personal/bundler-gem-bug/vendor/bundle/ruby/2.7.0/cache/rake-13.0.1.gem
An error occurred while installing rake (13.0.1), and Bundler cannot continue.
Make sure that `gem install rake -v '13.0.1' --source 'https://rubygems.org/'` succeeds before bundling.

In Gemfile:
  rake
```

**Comment by deivid-rodriguez** [comment]
Ok, thanks for confirming. I'll have a look!

**Comment by deivid-rodriguez** [comment]
Hello @captn3m0!

I had a look at your issue. As far as I could see, there's three problems:

* The first problem is specific to using a global gem cache. When you use a global gem cache, the `.gem` file in the global gem cache is first copied to the local cache, and the copied file is used. I'm not sure why this is necessary, if we used the file in the global cache directly, the error message would point to the right location.

* The second problem is not related to using a global gem cache. It's due to https://github.com/rubygems/bundler/pull/6010. If we leave the corrupt file in there, it will pollute the cache, making it impossible to retry the download. So we remove it. But then the error message is always going to point to a non existing file. I guess we could copy the corrupt file to a temporary location for inspection and show that in the error message.

* The third problem is that we don't special case empty files. Not sure if this is something that happens in real life, and worth fixing.

Thoughts?

**Comment by captn3m0** [comment]
Hmm, on (3) - looks like the empty files are anyway getting noticed and deleted from the local copied version - so I guess that behaviour doesn't need changing.

I guess solving (1) solves everything here? I don't really use bundler that much these days, but I've often been annoyed by this. From my understanding:

1. If I have a corrupt gem in my global cache.
2. It gets copied to the local cache.
3. bundler attempts an install, which fails.
4. Since the cache is clearly corrupt, we delete the file, but this is just the local cache.
5. Then we give an error message about the deleted gem file being corrupt.

Everything from 1-4 seems reasonable to me. Maybe another fix might be to just have a log message telling the user that the global cache was used (and the path to the original file).

The biggest issue when I faced this was not being able to figure out what was broken. Having a log pointing to the original gem cache (or just a `Note: Global cache is being used from ~/.bundler/cache. Gem file paths may be reported incorrect` at the top of the install log) would have helped.

**Comment by deivid-rodriguez** [comment]
I see. So essentially the problem here is again cache pollution like in https://github.com/rubygems/bundler/pull/6010, and as long as we also delete the corrupt file in the global cache, it should be fine and work upon retry. At least, it wouldn't be any worse than not using local cache. In both case the error message will still point to a deleted file, but retrying the download should work.

I'm still not sure why we keep both locations up to date when using global cache and I believe it would be easier to not care about the local cache at all when using a global cache.

**Comment by captn3m0** [comment]
Hmm, on (3) - looks like the empty files are anyway getting noticed and deleted from the local copied version - so I guess that behaviour doesn't need changing.

I guess solving (1) solves everything here? I don't really use bundler that much these days, but I've often been annoyed by this. From my understanding:

1. If I have a corrupt gem in my global cache.
2. It gets copied to the local cache.
3. bundler attempts an install, which fails.
4. Since the cache is clearly corrupt, we delete the file, but this is just the local cache.
5. Then we give an error message about the deleted gem file being corrupt.

Everything from 1-4 seems reasonable to me. Maybe another fix might be to just have a log message telling the user that the global cache was used (and the path to the original file).

The biggest issue when I faced this was not being able to figure out what was broken. Having a log pointing to the original gem cache (or just a `Note: Global cache is being used from ~/.bundler/cache. Gem file paths may be reported incorrect` at the top of the install log) would have helped.

**Comment by deivid-rodriguez** [comment]
Did you mean to post the same message again?

**Comment by captn3m0** [comment]
No, GitHub wasn't showing it to me even after a few refreshes. Deleted now.

**Comment by captn3m0** [comment]
Is there any feedback needed from my end here? I don't think I can add anything more, but I'm still facing this.
